### ISSUE TITLE
Generate specific functions for gtk.Builder getters

### ISSUE THREAD

**Issue by csoriano1618** [gotk3/gotk3#331]
One nice thing of Go vs C is that we can make sure type enforcement is done well tm. It would be ideal to have type specific getters for gtk.Builder, so you would have something like builder.GetEntry instead of builder.GetObject. This will also make sure the result is already casted.

This idea is nothing new, rust-gtk does it for signals, so you always connect to a specific signal instead of the more abstract object.Connect().

This idea can be implemented with generators, there is a [prototype](https://gitlab.gnome.org/csoriano/weeport/tree/master/internal/builder) done by @didrocks of this idea extending gtk.Builder. The result of the generator is something like:

```
// GetButton returns the matching *gtk.Button corresponding to name
func (b Builder) GetButton(name string) (*gtk.Button, error) {
	iobj, err := b.Builder.GetObject(name)
	if err != nil {
		return nil, xerrors.Errorf("couldn't find '%s' in ui file: %v", name, err)
	}

	widget, ok := iobj.(*gtk.Button)
	if !ok {
		return nil, xerrors.Errorf("'%s' isn't a *gtk.Button", name)
	}

	return widget, nil
}
```
And the usage as simple as
```
b := gtk.BuilderNew()
t := b.GetButton("buttonID")
```

So I would like to know what you think of this concept, and if something similar could be implemented.

**Comment by Bios-Marcel** [comment]
About connecting to specific signals, vala does this too and I like it, it's much better than string based APIs.  I generally dislike every API that is string based. 

Generally I like the specific getters that already do the casting for you as well. However, whether those are generated or not isn't really important I assume. I'd generally like to avoid auto generating that stuff, because I see no big benefit, writing those boilerplate functions once doesn't consume too much time.

**Comment by csoriano1618** [comment]
Once you have signals, types, etc. and things like gtk4, etc. it's gonna be a life changer to autogenerate based on a simple list. And actually, once (if) the binding starts using the gir, generating stuff is gonna be the default and not the opposite.

**Comment by Bios-Marcel** [comment]
I don't think that I am currently in a position where I can judge whether generating big parts of the bindings is possible at all. I just lack the knowledge. But it would be quite useful indeed, you are right.

I heard about rust bindings for some libraries being completely auto generated. I am not sure whether I missunderstood that or not, but that'd be lovely. However, I don't think if that's possible with go and how much work it requires.

**Comment by csoriano1618** [comment]
That's correct, other bindings such as Rust, JS and Python are autogenerated by using [gir](https://gi.readthedocs.io/en/latest/), it's how bindings keep being reasonable sane to maintain. Otherwise they would be a nightmare to keep up to date.

FWIW none of them started by autogenerating, I think that's the usual life cycle of a binding. So it's not that the current Go binding started in the wrong foot, it's just that it's still in a early life cycle :-).

> However, I don't know if that's possible with go and how much work it would require.

I believe it's actually the opposite, I think go is the best positioned in this regard. Go has a powerful generator story, compared to what other bindings need to do the Go one is probably the one that can be made with less effort.

It's still work though. Based on this past month that I tried the bindings, and considering the go generator system, I have the feeling it's a great time to start transitioning into autogenerated bindings.

**Comment by Bios-Marcel** [comment]
Well I know you can make simple stuff, but what about more complex cases, where you need actual logic instead of just delegation function calls or so. I have never done anything like that, so my concern might be wrong though. I would be happy if someone would prove me completely wrong :D

**Comment by andre-hub** [comment]
if someone would take the time to write this generator for certain parts it might make sense. However, there are cases where there is some more logic built in, even if it is only a loop or a cast. These areas can also be mapped with a generator, but also cases need time for implementation.

**Comment by andre-hub** [comment]
https://github.com/gotk3/gotk3/wiki/gotk3-without-generator

we can begin with an todo for the change to generator based wapper
